### PR TITLE
rtrlib: ability to pass configured socket in tcp/ssh attribute

### PR DESCRIPTION
**Pull request by pguibert6WIND**
a new attribute socket is given in tr_ssh_config and tr_tcp_config
structures. as the structures are visible for configuration by the
caller, it becomes possible to create early the socket that will be
used later by rtrlib. some scenarios like vrf usage can be possible.
for instance, the socket is first created in the appropriate vrf, then
submitted to rtrlib that uses it to connect to the target.
Also, tr_ident_fp is modified so as to return the configured socket
ctxt. this is helpful to retrieve the context.

Signed-off-by: Philippe Guibert <philippe.guibert@6wind.com>

### CODE REVIEW

**Comment by mroethke** [comment]
This is a good start, but it won't work.
The socket is closed in `tr_tcp_close` and I assume `ssh_channel_close` does the same.
There needs to be a way to request a new one, I suggest a callback with the following signature:
```c
int new_socket(void *data);
```
where data points to caller provided data, which should be stored in the config struct too.
This assumes the callback does not call `connect` on it's own, I don't know if that would be useful.

And I don't think it makes sense to put the sockets fd number in the ident string.

**Comment by pguibert6WIND** [comment]
just a quick update to see if it conforms to your counter proposal.

File: rtrlib/transport/ssh/ssh_transport.c [hunk 1]
```diff
@@ -50,6 +50,7 @@ int tr_ssh_open(void *socket)
 {
 	struct tr_ssh_socket *ssh_socket = socket;
 	const struct tr_ssh_config *config = &ssh_socket->config;
+	int fd;
```
Comment (mroethke) [inline rtrlib/transport/ssh/ssh_transport.c]: please move the declaration into the if, as it is not used outside of it.

File: rtrlib/transport/ssh/ssh_transport.c [hunk 2]
```diff
@@ -74,6 +75,11 @@ int tr_ssh_open(void *socket)
 
 	if (config->client_privkey_path)
 		ssh_options_set(ssh_socket->session, SSH_OPTIONS_IDENTITY, config->client_privkey_path);
+	if (config->new_socket && config->opaque_info) {
```
Comment (mroethke) [inline rtrlib/transport/ssh/ssh_transport.c]: I'm not sure it makes sense to require new_socket and opaque_info to be set, some might not care about any context information.

File: rtrlib/transport/tcp/tcp_transport.c [hunk 3]
```diff
@@ -48,6 +48,7 @@ int tr_tcp_open(void *tr_socket)
 {
 	int rtval = TR_ERROR;
 	struct tr_tcp_socket *tcp_socket = tr_socket;
+	const struct tr_tcp_config *config = &tcp_socket->config;
```
Comment (mroethke) [inline rtrlib/transport/tcp/tcp_transport.c]: I'm not against adding a variable as a shortcut, but right now the use in this function is inconsistent `tcp_socket->config` is still used in several places.
Keeping the `config` variable in this PR would be fine by me, but I would create a separate commit for it's addition and use in the function.

**Comment by pguibert6WIND** [comment]
Marcel,
do you have an idea on what is wrong with the syntax check:
```
+++ b/rtrlib/transport/tcp/tcp_transport.c
@@ -65,7 +65,7 @@ int tr_tcp_open(void *tr_socket)

	if (config->new_socket) {
              config->new_socket (...)
       }
```
actually, I dont know why, test fail because of that piece of code ( call is done despite new_socket is supposed to be null).

File: rtrlib/transport/tcp/tcp_transport.h [hunk 4]
```diff
@@ -28,11 +28,17 @@
...
+ * @param opaque_info Information to pass to callback function
+ *	  in charge of retrieving socket
+ * @param new_socket(void *opaque_info) callback routine, that
+ *	  if defined, will obtain the socket used for connection
  */
 struct tr_tcp_config {
 	char *host;
 	char *port;
 	char *bindaddr;
+	void *opaque_info;
```
Comment (mroethke) [inline rtrlib/transport/tcp/tcp_transport.h]: In other places we call this just `data`, might be better for consistency.

**Comment by mroethke** [comment]
Some additional info:

The config struct is copied in `tr_{tcp,ssh}_init`, which must be extended for the new struct fields.

**Comment by pguibert6WIND** [comment]
Hi @mroethke ,
yes I realized that the crash I had was due to the non initialisation of tr[tcp/ssh]_init

**Comment by pguibert6WIND** [comment]
Hi @mroethke ,
I could use that branch and make it work with associated patch from frr.
so from the operational perspective, this is ok from my side.

File: rtrlib/transport/ssh/ssh_transport.c [hunk 5]
```diff
@@ -74,6 +74,17 @@ int tr_ssh_open(void *socket)
 
 	if (config->client_privkey_path)
 		ssh_options_set(ssh_socket->session, SSH_OPTIONS_IDENTITY, config->client_privkey_path);
+	if (config->new_socket) {
+		int fd;
+
+		fd = config->new_socket(config->data);
+		if (fd > 0)
```
Comment (mroethke) [inline rtrlib/transport/ssh/ssh_transport.c]: same as with tcp

**Comment by waehlisch** [comment]
> But since this changes the API I would like to have another opinion.
>
maybe @fho or @nbars have also opinions.

File: rtrlib/transport/tcp/tcp_transport.h [hunk 6]
```diff
@@ -28,11 +28,18 @@
  * @param bindaddr Hostname or IP address to connect from. NULL for
  *		   determination by OS.
  * to use the source address of the system's default route to the server
+ * @param data Information to pass to callback function
+ *	  in charge of retrieving socket
+ * @param new_socket(void *opaque_info) callback routine, that
+ *	  if defined, will obtain the socket used for connection.
+ *	  the socket will be connected and unbound if success
```
[the remark this answers is not in the document]
Comment (pguibert6WIND) [reply rtrlib/transport/tcp/tcp_transport.h]: I was looking at how code in the frr code looked like.
I rephrased it.

File: rtrlib/transport/tcp/tcp_transport.c [hunk 7]
```diff
@@ -48,50 +48,65 @@ int tr_tcp_open(void *tr_socket)
...
-		goto end;
+	if (config->new_socket) {
+		tcp_socket->socket = (*config->new_socket)(config->data);
+		if (tcp_socket->socket <= 0) {
+			TCP_DBG("Couldn't establish TCP connection, %s",
+				tcp_socket, strerror(errno));
+			goto end;
+		}
 	}
+	if (tcp_socket->socket <= 0) {
```
[the remark this answers is not in the document]
Comment (pguibert6WIND) [reply rtrlib/transport/tcp/tcp_transport.c]: ok, I will keep < 0 , then.

File: rtrlib/transport/tcp/tcp_transport.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (mroethke) [reply rtrlib/transport/tcp/tcp_transport.h]: I still don't like the wording.

How about:
> Pointer to function that is called every time a new connections is made. The returned socket is expected to be ready for use (e.g. in state established) and must use a reliably stream-oriented transport.

That would be more in line with the docstrings of other callback functions in rtrlib.

[the remark this answers is not in the document]
Comment (pguibert6WIND) [reply rtrlib/transport/tcp/tcp_transport.h]: ok.

**Comment by nbars** [comment]
Hi,

the added fields are in no way related to the semantics of the attributes that ```tr_tcp_init``` currently has. The same is true for ```tr_ssh_config```. As indicated by the duplicated code added to both structs, it makes sense (in my opinion) to refactor these changes into a new API function. Since your changes are not related to TCP/SSH in particular, I think the right place would be in ```transport.c```.
Consequently, no breaking changes to the old API and ~clean~ separation of concerns.

If I got something wrong, please tell me. Furthermore, in case you keep your changes, the documentation must be improved to make the mutual exclusiveness of the old and new fields clear.

**Comment by pguibert6WIND** [comment]
Hi @nbars,
thanks for feedback.
I will think of your proposal.
What I can say is that with that state of code, I am not sure ssh can work ( how can I test ssh). reason is that I am not sure a connected socket would work with ssh.
Practically, if I call connect() once, then ssh-dev will call a second time connect().. what happens ?

**Comment by mroethke** [comment]
For testing you could either use [gortr](https://github.com/cloudflare/gortr) or put something like 
> Subsystem       rpki-rtr socat STDIO TCP:localhost:3323

in your sshd config. Where localhost:3323 is where the rtr server listens on TCP.

I did some testing with the current version and ssh works. Apparently libssh expects a connected socket even though that is not clear in the documentation.

@nbars 
There needs to be transport specific code anyway and I would prefer one place to configure a socket completely to to have it scattered over several function calls.
But you are right about the documentation.

**Comment by pguibert6WIND** [comment]
Hi,

| I did some testing with the current version and ssh works. 

It should be tested with https://github.com/FRRouting/frr/pull/5015.
I plan to test it next week, unless someone did not do it before

| Apparently libssh expects a connected socket even though that is not clear in the documentation.

my doubt is about getaddrinfo() being called in ssh-dev layer, whereas it is already called in frr.
I fear that the address settings may make frr+rtrlib in ssh mode buggy.

where could I add some more documentation to clarify better ? in specific tcp/ssh headers ?

**Comment by mroethke** [comment]
> my doubt is about getaddrinfo() being called in ssh-dev layer, whereas it is already called in frr.
> I fear that the address settings may make frr+rtrlib in ssh mode buggy.

Why would that happen if libssh expects a connected socket?

> where could I add some more documentation to clarify better ? in specific tcp/ssh headers?

Yes

**Comment by pguibert6WIND** [comment]
> > my doubt is about getaddrinfo() being called in ssh-dev layer, whereas it is already called in frr.
> > I fear that the address settings may make frr+rtrlib in ssh mode buggy.
> 
> Why would that happen if libssh expects a connected socket?
> 
I need to look deeper on that. I dont think libssh exepcts a connected socket first.
then if on the call back, I create socket, and prepare it(  getaddrinfo()) 
then if on ssh-dev, the socket is prepared again ( getaddrinfo()), then connect() is called.

I think this will not work.

**Comment by pguibert6WIND** [comment]
ok.
for the established status for ssh, you are right.
actually, in ssh-dev code, it looks as if it was already connected.
we are on the first if (), and we do not perform ssh_connect.

```
  if (session->opts.fd != SSH_INVALID_SOCKET) {
    session->session_state=SSH_SESSION_STATE_SOCKET_CONNECTED;
    ssh_socket_set_fd(session->socket, session->opts.fd);
    ret=SSH_OK;
#ifndef _WIN32
  } else if (session->opts.ProxyCommand != NULL){
    ret = ssh_socket_connect_proxycommand(session->socket,
                                          session->opts.ProxyCommand);
#endif
  } else {
    ret=ssh_socket_connect(session->socket,
                           session->opts.host,
                           session->opts.port > 0 ? session->opts.port : 22,
                           session->opts.bindaddr);
  }
  if (ret == SSH_ERROR) {
    return SSH_ERROR;
  }
```

**Comment by pguibert6WIND** [comment]
ok,
just an update to fix travis.
I did look at documentation in headers, and it looks ok to me ( connect is mentioned) for both tcp/ssh.
I did not test as my knowledge about socat is very limited, and have few time to look at it. so I tend to more rely on your opinion Marcel, if you said you tested it, or if you re willing to help me more about socat.
this said, I think the priority is the review.

**Comment by mroethke** [comment]
> I did look at documentation in headers, and it looks ok to me ( connect is mentioned) for both tcp/ssh.

Yes, but it is not mentioned that when `new_socket` is set `host`, `port`, and `bindaddr` are not used at all (except for logging output).

> I did not test as my knowledge about socat is very limited, and have few time to look at it. so I tend to more rely on your opinion Marcel, if you said you tested it, or if you re willing to help me more about socat.

As I wrote above:
Just put 

     Subsystem rpki-rtr socat STDIO TCP:$rtr_server_host:$rtr_server_port

in you sshd config, you may use our public rtr server.
It listens on `rpki-validator.realmv6.org:8283`.

**Comment by pguibert6WIND** [comment]
Hi @mroethke , long time no discussed with you.
so, a few more questions on how to set up the ssh connection:
```
 Subsystem rpki-rtr socat STDIO TCP:141.22.28.252:8283
```
and regarding frr configuratio, this will be something like that, right ?
```
vrf vrf0                                                           
 rpki                                                              
  rpki cache 141.22.28.252 8283 root /root/.ssh/id_rsa /root/.ssh/id_rsa.pub preference 2
  exit                                                             
 exit-vrf 
```
I got folllowing message
```
(2020/01/03 15:36:05:672771): RTR_MGR: rtr_mgr_start()
(2020/01/03 15:36:05:675423): RTR Socket: State: RTR_CONNECTING
(2020/01/03 15:36:05:676714): SSH Transport(root@141.22.28.252:8283): tr_ssh_init: opening SSH connection failed
(2020/01/03 15:36:05:677980): SSH Transport(root@141.22.28.252:8283): Socket closed
(2020/01/03 15:36:05:679596): RTR_MGR: Group(3) status changed to: RTR_MGR_ERROR
(2020/01/03 15:36:05:680141): RTR_MGR: No other inactive groups found
(2020/01/03 15:36:05:681245): RTR Socket: State: RTR_ERROR_TRANSPORT
(2020/01/03 15:36:05:682702): SSH Transport(root@141.22.28.252:8283): Socket closed
(2020/01/03 15:36:05:683393): RTR_MGR: Group(3) status changed to: RTR_MGR_ERROR
(2020/01/03 15:36:05:684066): RTR Socket: Waiting 600
```
I assume something went wrong, right?

**Comment by pguibert6WIND** [comment]
added comments about ip, port, bindaddr not used when new_socket() is called.

**Comment by mroethke** [comment]
Sorry for the late answer.

You have to configure the ssh host, where you configured the rtr subsystem, as the rtr server in frr.

**Comment by pguibert6WIND** [comment]
a change in the frr config with a standard 7.0 based release without all rpki changes:
```
rpki cache 127.0.0.1 22 root /root/.ssh/id_rsa /root/.ssh/id_rsa.pub  /etc/ssh/ssh_host_rsa_key.pub preference 5
```
along with ssh service change ( based on ubuntu 1804 distrib)
```
 Subsystem rpki-rtr socat STDIO TCP:141.22.28.252:8283
```
I fell on an unknown error where somehow, the ssh daemon of localhost did not finalise the ssh session ( despite the fact that If I Try to connect with ssh -i <key> root@localhost, it works.
```
Jan 31 18:56:25 ubuntu1804hwe sshd[28478]: debug3: fd 5 is not O_NONBLOCK
Jan 31 18:56:25 ubuntu1804hwe sshd[28478]: debug1: Forked child 29630.
Jan 31 18:56:25 ubuntu1804hwe sshd[28478]: debug3: send_rexec_state: entering fd = 8 config len 548
Jan 31 18:56:25 ubuntu1804hwe sshd[28478]: debug3: ssh_msg_send: type 0
Jan 31 18:56:25 ubuntu1804hwe sshd[28478]: debug3: send_rexec_state: done
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug3: oom_adjust_restore
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug1: Set /proc/self/oom_score_adj to 0
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug1: rexec start in 5 out 5 newsock 5 pipe 7 sock 8
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug1: inetd sockets after dupping: 3, 3
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: Connection from 127.0.0.1 port 43594 on 127.0.0.1 port 22
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug1: Client protocol version 2.0; client software version libssh_0.8.90
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug1: no match: libssh_0.8.90
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug1: Local version string SSH-2.0-OpenSSH_7.6p1 Ubuntu-4ubuntu0.3
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug2: fd 3 setting O_NONBLOCK
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug3: ssh_sandbox_init: preparing seccomp filter sandbox
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug2: Network child is on pid 29631
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug3: preauth child monitor started
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug3: privsep user:group 106:65534 [preauth]
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug1: permanently_set_uid: 106/65534 [preauth]
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug3: ssh_sandbox_child: setting PR_SET_NO_NEW_PRIVS [preauth]
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug3: ssh_sandbox_child: attaching seccomp filter program [preauth]
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug1: list_hostkey_types: ssh-rsa,rsa-sha2-512,rsa-sha2-256 [preauth]
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug3: send packet: type 20 [preauth]
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug1: SSH2_MSG_KEXINIT sent [preauth]
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug3: receive packet: type 20 [preauth]
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug1: SSH2_MSG_KEXINIT received [preauth]
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug2: local server KEXINIT proposal [preauth]
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug2: KEX algorithms: curve25519-sha256,curve25519-sha256@libssh.org,ecdh-sha2-nistp256,ecdh-sha2-nistp384,ecdh-sha2-nistp521,diffie-hellman-grou
,diffie-hellman-group18-sha512,diffie-hellman-group14-sha256,diffie-hellman-group14-sha1 [preauth]

Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug2: host key algorithms: ssh-rsa,rsa-sha2-512,rsa-sha2-256 [preauth]
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug2: ciphers ctos: chacha20-poly1305@openssh.com,aes128-ctr,aes192-ctr,aes256-ctr,aes128-gcm@openssh.com,aes256-gcm@openssh.com [preauth]
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug2: ciphers stoc: chacha20-poly1305@openssh.com,aes128-ctr,aes192-ctr,aes256-ctr,aes128-gcm@openssh.com,aes256-gcm@openssh.com [preauth]
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug2: MACs ctos: umac-64-etm@openssh.com,umac-128-etm@openssh.com,hmac-sha2-256-etm@openssh.com,hmac-sha2-512-etm@openssh.com,hmac-sha1-etm@opens
,hmac-sha2-256,hmac-sha2-512,hmac-sha1 [preauth]
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug2: MACs stoc: umac-64-etm@openssh.com,umac-128-etm@openssh.com,hmac-sha2-256-etm@openssh.com,hmac-sha2-512-etm@openssh.com,hmac-sha1-etm@opens
,hmac-sha2-256,hmac-sha2-512,hmac-sha1 [preauth]
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug2: compression ctos: none,zlib@openssh.com [preauth]
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug2: compression stoc: none,zlib@openssh.com [preauth]
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug2: languages ctos:  [preauth]
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug2: languages stoc:  [preauth]
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug2: first_kex_follows 0  [preauth]
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug2: reserved 0  [preauth]
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug2: peer client KEXINIT proposal [preauth]
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug2: KEX algorithms: curve25519-sha256,curve25519-sha256@libssh.org,ecdh-sha2-nistp256,ecdh-sha2-nistp384,ecdh-sha2-nistp521,diffie-hellman-grou
-hellman-group-exchange-sha256,diffie-hellman-group14-sha1,diffie-hellman-group1-sha1,ext-info-c [preauth]


Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug2: host key algorithms: ssh-ed25519,ecdsa-sha2-nistp521,ecdsa-sha2-nistp384,ecdsa-sha2-nistp256,rsa-sha2-512,rsa-sha2-256,ssh-rsa,ssh-dss [pre
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug2: ciphers ctos: aes256-gcm@openssh.com,aes128-gcm@openssh.com,aes256-ctr,aes192-ctr,aes128-ctr,aes256-cbc,aes192-cbc,aes128-cbc,3des-cbc [pre
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug2: ciphers stoc: aes256-gcm@openssh.com,aes128-gcm@openssh.com,aes256-ctr,aes192-ctr,aes128-ctr,aes256-cbc,aes192-cbc,aes128-cbc,3des-cbc [pre
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug2: MACs ctos: hmac-sha2-256-etm@openssh.com,hmac-sha2-512-etm@openssh.com,hmac-sha1-etm@openssh.com,hmac-sha2-256,hmac-sha2-512,hmac-sha1 [pre
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug2: MACs stoc: hmac-sha2-256-etm@openssh.com,hmac-sha2-512-etm@openssh.com,hmac-sha1-etm@openssh.com,hmac-sha2-256,hmac-sha2-512,hmac-sha1 [pre
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug2: compression ctos: none [preauth]
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug2: compression stoc: none [preauth]
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug2: languages ctos:  [preauth]
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug2: languages stoc:  [preauth]
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug2: first_kex_follows 0  [preauth]
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug2: reserved 0  [preauth]
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug1: kex: algorithm: curve25519-sha256 [preauth]
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug1: kex: host key algorithm: rsa-sha2-512 [preauth]
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug1: kex: client->server cipher: aes256-gcm@openssh.com MAC: <implicit> compression: none [preauth]
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug1: kex: server->client cipher: aes256-gcm@openssh.com MAC: <implicit> compression: none [preauth]
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug1: expecting SSH2_MSG_KEX_ECDH_INIT [preauth]
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug3: receive packet: type 30 [preauth]
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug3: mm_key_sign entering [preauth]
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug3: mm_request_send entering: type 6 [preauth]
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug3: mm_key_sign: waiting for MONITOR_ANS_SIGN [preauth]
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug3: mm_request_receive_expect entering: type 7 [preauth]
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug3: mm_request_receive entering [preauth]
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug3: mm_request_receive entering
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug3: monitor_read: checking request 6
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug3: mm_answer_sign
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug3: mm_answer_sign: hostkey proof signature 0x564dc39267c0(276)
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug3: mm_request_send entering: type 7
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug2: monitor_read: 6 used once, disabling now
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug3: send packet: type 31 [preauth]
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug3: send packet: type 21 [preauth]
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug2: set_newkeys: mode 1 [preauth]
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug1: rekey after 4294967296 blocks [preauth]
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug1: SSH2_MSG_NEWKEYS sent [preauth]
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug1: expecting SSH2_MSG_NEWKEYS [preauth]
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug3: send packet: type 7 [preauth]
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug3: receive packet: type 21 [preauth]
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug1: SSH2_MSG_NEWKEYS received [preauth]
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug2: set_newkeys: mode 0 [preauth]
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug1: rekey after 4294967296 blocks [preauth]
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug1: KEX done [preauth]
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: Connection closed by 127.0.0.1 port 43594 [preauth]
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug1: do_cleanup [preauth]
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug3: PAM: sshpam_thread_cleanup entering [preauth]
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug1: monitor_read_log: child log fd closed
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug3: mm_request_receive entering
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug1: do_cleanup
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug3: PAM: sshpam_thread_cleanup entering
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug1: Killing privsep child 29631
Jan 31 18:56:25 ubuntu1804hwe sshd[29630]: debug1: audit_event: unhandled event 12
```
@mroethke , do you mind if I ask you more information about the env settings you are using to simulate the SSH environment, please ?

**Comment by mroethke** [comment]
Did you install socat?

**Comment by pguibert6WIND** [comment]
socat version 1.7.3.2 on Apr  4 2018 10:06:49

**Comment by pguibert6WIND** [comment]
looks like my ssh settings are incirrect 
```
dut-vm(config)# (2020/02/03 10:04:00:719785): SSH Transport(root@127.0.0.1:22): tr_ssh_init: Wrong hostkey
```

**Comment by pguibert6WIND** [comment]
ok, the tip was, instead of using the public key directly, (`/etc/ssh/ssh_host_rsa_key.pub` file ), the expectation is to have `/root/.ssh/known_hosts` as server public key.

**Comment by pguibert6WIND** [comment]
Hi @mroethke,
so finally I could make a setup and test ssh via local ssh server and socat subsystem procedure.
I made it work with:
- rtrlib with initial upstream code
- rtrlib with ssh on default vrf (with frr branch https://github.com/FRRouting/frr/pull/5015)
- rtrlib with ssh on custom vrf (with frr branch https://github.com/FRRouting/frr/pull/5015)

config example used:
```
rpki
rpki cache 127.0.0.1 22 root /root/.ssh/id_rsa /root/.ssh/id_rsa.pub  /root/.ssh/known_hosts preference 5
exit
```

thanks to let me know if there are still pending points to discuss, or can this pr be unblocked, now that SSH has been tested.
thanks,

**Comment by pguibert6WIND** [comment]
Hi @mroethke ,
thanks again for the help on this ticket.
now I am waiting for your positive feedback on this PR.
I would really like to go ahead with new FRR RPKI stuff.
thanks.

**Comment by waehlisch** [comment]
ping @mroethke @nbars

**Comment by mroethke** [comment]
I think it looks good now.

But would still like to have another opinion.

**Comment by pguibert6WIND** [comment]
thanks Marcel, I am looking forward to get news on this pr

**Comment by pguibert6WIND** [comment]
Hi,
are there any feedbacks there, please?

**Comment by waehlisch** [comment]
ping @mroethke

**Comment by mroethke** [comment]
coverity pointed out to me, that this breaks the rtrclient. Because the new fields are not initialized. Same for current version of frr.

Fixing this in rtrclient is as simple as this:
``` patch
diff --git a/tools/rtrclient.c b/tools/rtrclient.c
index 3b89e11..0fb7749 100644
--- a/tools/rtrclient.c
+++ b/tools/rtrclient.c
@@ -755,9 +755,9 @@ static void init_sockets(void)
 {
 	for (size_t i = 0; i < socket_count; ++i) {
 		struct socket_config *config = socket_config[i];
-		struct tr_tcp_config tcp_config;
+		struct tr_tcp_config tcp_config = {};
 #ifdef RTRLIB_HAVE_LIBSSH
-		struct tr_ssh_config ssh_config;
+		struct tr_ssh_config ssh_config = {};
 #endif
 
 		switch (config->type) {
```

Please add this, than I will merge.

I will also prepare fixes for supported version of frr asap.

**Comment by pguibert6WIND** [comment]
strange: on https://ci1.netdef.org/browse/RPKI-RTRLIBPR-FEDORA29AMD64-275/log,
```
13-Mar-2020 01:01:05 | 4 packages and 0 specfiles checked; 4 errors, 2 warnings.
-- | --
13-Mar-2020 01:01:05 | debian package build (debuild) failed
```

is there a way to relaunch CI ?

**Comment by pguibert6WIND** [comment]
ci:rerun

**Comment by mroethke** [comment]
I just merged and pushed this manually.

Thanks!

**Comment by pguibert6WIND** [comment]
Hi @waehlisch , who could help on having an official delivery of rtrlib ?
This pull request is a requisite for https://github.com/FRRouting/frr/pull/5015 that has been reverted because rtrlib was not up to date with distribs.

**Comment by waehlisch** [comment]
@pguibert6WIND rtrlib is available as an official Debian package (https://packages.debian.org/search?keywords=librtr-dev) -- you mean we need to update the package?

**Comment by pguibert6WIND** [comment]
@waehlisch , that is it. updating the package would be very nice. I know it takes time. I dont know who to contact and how to package to different distros.

**Comment by waehlisch** [comment]
last time @lukas227 helped us. ping. but, librtr-0.7 is already part of stable (see Version widget https://tracker.debian.org/pkg/librtr).

what we need to do is to create a new RTRlib release. this will then be adapted in Debian.

**Comment by lukas227** [comment]
Happy update the package in Debian once you've created a new release.

**Pull request closed without merging**